### PR TITLE
Catlin: Report logs back to the PR after execution

### PR DESCRIPTION
**Pull request by vinamra28**
# Changes

As of now catlin is running on PRs of catalog and validate them as per
TEP-0003 but there are some warnings which are produced by catlin which
maybe left un-noticed by the PR author as catlin check doesn't fails.

This PR adds a script which will search whether catlin has produced any
warning or error and if found will post the output as comment on the PR.

/kind feature

Signed-off-by: vinamra28 <vinjain@redhat.com>
# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: tekton/ci/jobs/tekton-catlin-lint.yaml [hunk 1]
```diff
@@ -80,3 +97,21 @@ spec:
...
+        - input: $(tasks.lint-catalog.results.catlin-result)
+          operator: "notin"
+          values: ["no"]
+      params:
+        - name: COMMENT
+          value: $(tasks.lint-catalog.results.changed-tasks)
+        - name: GITHUB_TOKEN_SECRET_NAME
+          value: bot-token-github
+        - name: GITHUB_TOKEN_SECRET_KEY
+          value: bot-token
```
Comment (vinamra28) [inline tekton/ci/jobs/tekton-catlin-lint.yaml:113]: is there a secret setup which contains the `base64` personal access token? If yes, then please help me with the correct values that I can substitute over here :slightly_smiling_face:

**Review by vinamra28 — COMMENTED** [review]
also please do tell where to document that [github-add-comment](https://hub.tekton.dev/tekton/task/github-add-comment) task is being used from tektoncd/catalog :sweat_smile:

File: tekton/ci/jobs/tekton-catlin-lint.yaml [hunk 2]
```diff
@@ -34,11 +37,33 @@ spec:
       image: gcr.io/tekton-releases/dogfooding/catlin:latest
       workingDir: $(resources.inputs.source.path)
       script: |
+        set +e
```
Comment (vinamra28) [inline tekton/ci/jobs/tekton-catlin-lint.yaml:40]: I just realized that the job is going to always pass with this `set +e`. Will fix this tomorrow :sweat_smile:

**Comment by vinamra28** [comment]
/hold cancel
but this requires https://github.com/tektoncd/catalog/pull/664 to be merged first :sweat_smile:

**Comment by vinamra28** [comment]
@afrittoli since tektoncd/catalog#664 is now merged, can you please tell where to document that I need this task from upstream catalog and how to move forward with this PR?

File: tekton/ci/jobs/tekton-catlin-lint.yaml (hunk 1 above)
Comment (afrittoli) [reply tekton/ci/jobs/tekton-catlin-lint.yaml:116]: The values you have in there are already ok

File: tekton/ci/jobs/tekton-catlin-lint.yaml [hunk 3]
```diff
@@ -34,11 +34,32 @@ spec:
...
+        catlin validate $(cat $(workspaces.store-changed-files.path)/changed-files.txt) | tee -a $(workspaces.store-changed-files.path)/catlin.txt
+        echo '```' >> $(workspaces.store-changed-files.path)/catlin.txt
+
+        # checking if there is any ERROR or WARN produces by catlin
+        isWarning=$(cat  $(workspaces.store-changed-files.path)/catlin.txt | grep -c "WARN")
+        isError=$(cat  $(workspaces.store-changed-files.path)/catlin.txt | grep -c "ERROR")
+
+        # if there is any ERROR or WARN then post the comment else skip
+        [[ $isWarning -eq 0 ]] && [[ $isError -eq 1 ]] && \
+        echo -n "" > $(workspaces.store-changed-files.path)/catlin.txt
```
Comment (afrittoli) [inline tekton/ci/jobs/tekton-catlin-lint.yaml:59]: I don't really understand the logic here, perhaps it should be `$isError -eq 0`? 

BTW, it looks like the `post-comment` will be executed regardless, only with an empty comment?
Hopefully the github API is smart enough to not post an empty comment?

A nice solution would be to have a result and use `when` to skip the comment if not needed. Unfortunately `when` expressions are not available in `finally` yet (https://github.com/tektoncd/community/pull/240).

**Review by afrittoli — COMMENTED** [review]
Thank you!
It looks mostly ok, but I think there may be a problem in the script logic.

File: tekton/ci/jobs/tekton-catlin-lint.yaml (hunk 3 above)
Comment (vinamra28) [reply tekton/ci/jobs/tekton-catlin-lint.yaml:59]: > I don't really understand the logic here, perhaps it should be `$isError -eq 0`?
> 
oh yes :sweat_smile:, it should be `$isError -eq 0`. Will fix this :sweat_smile: 
> BTW, it looks like the `post-comment` will be executed regardless, only with an empty comment?
> Hopefully the github API is smart enough to not post an empty comment?
> 
yeah when there is an empty comment then no comment would be posted on PR

> A nice solution would be to have a result and use `when` to skip the comment if not needed. Unfortunately `when` expressions are not available in `finally` yet ([tektoncd/community#240](https://github.com/tektoncd/community/pull/240)).

Comment (vinamra28) [reply tekton/ci/jobs/tekton-catlin-lint.yaml]: btw the following script will produce the output as 

----
**Catlin Output**
```
FILE: task/black/0.1/black.yaml
WARN : Step "format-python-code" uses image "$(params.image)" that contains variables; skipping validation
FILE: task/yaml-lint/0.1/yaml-lint.yaml
```

File: tekton/ci/jobs/tekton-catlin-lint.yaml [hunk 4]
```diff
@@ -40,11 +40,33 @@ spec:
...
+        # creating a file which will contain the final formatted output
+        # which needs to be added as a comment(if any)
+        echo '**Catlin Output**' >> $(workspaces.store-changed-files.path)/catlin.txt
+        echo '```' >> $(workspaces.store-changed-files.path)/catlin.txt
+
+        # performing catlin validate
+        catlin validate $(cat $(workspaces.store-changed-files.path)/changed-files.txt) | tee -a $(workspaces.store-changed-files.path)/catlin.txt
+        echo '```' >> $(workspaces.store-changed-files.path)/catlin.txt
+
+        # checking if there is any ERROR or WARN produces by catlin
```
Comment (mpetason) [inline tekton/ci/jobs/tekton-catlin-lint.yaml:59]:
```suggestion
        # checking if there are any ERROR or WARN messages produced by catlin
```
messages or maybe log messages

File: tekton/ci/jobs/tekton-catlin-lint.yaml [hunk 5]
```diff
@@ -40,11 +40,33 @@ spec:
...
+        isWarning=$(cat  $(workspaces.store-changed-files.path)/catlin.txt | grep -c "WARN")
+        isError=$(cat  $(workspaces.store-changed-files.path)/catlin.txt | grep -c "ERROR")
+
+        # if there is no ERROR or WARN then add a empty string which will not
+        # add comment on Github PR
```
Comment (mpetason) [inline tekton/ci/jobs/tekton-catlin-lint.yaml:64]:
```suggestion
        # if there are no ERROR or WARN messages then add a empty string which will not
        # add a comment on the Github PR
```

**Review by vdemeester — APPROVED** [review]

**Pull request merged**